projects/rocm-core/rdhc/rdhc.py

```diff
@@ -0,0 +1,1795 @@
+                    self.logger.warning(f"!!! Library symlink {lib}->{rplib} ; pointing outside of ROCm library path {rocm_lib_path}.")
+                continue
+
+            stdout, stderr, ret_code = CommandExecutor.run_command(f"ldd {lib}", shell=True)
+            # Check if its not a dynamic library
+            if "not a dynamic executable" in stderr:
+                continue
+
+            if ret_code != 0:
+                missing_deps[lib] = f"Error running ldd: {stderr}"
+                continue
+
+            self.logger.debug(f"----Checking dependencies & link paths for {lib}...")
+            # Parse ldd output for any libraries that are not found in the system
+            # and for any linked libraries that are not in the ROCm library path and raise the warning.
+            for line in stdout.splitlines():
+                if "not found" in line:
+                    missing.append(line.strip())
+                elif "=>" in line:
+                    # Ex: "libamdhip64.so => /opt/rocm/lib/libamdhip64.so (0x00007f8c3c000000)"
+                    # Check if the library is outside of the ROCm library path
+                    parts = line.split("=>")
+                    if len(parts) > 1:
+                        dep_lib = parts[0].strip()
+                        dep_lib_path = parts[1].strip().split()[0]
+                        # dep_lib_path can be relative path, so we need to resolve it
+                        # Check if the link is relative or absolute
+                        if not os.path.isabs(dep_lib_path):
+                            # If it's relative, resolve it against the library path
+                            # normalize the path to remove any redundant separators
+                            dep_lib_path = os.path.normpath(os.path.join(os.path.dirname(lib), dep_lib_path))
+
+                        # check if the lib is a ROCm library, else # skip the check
+                        if dep_lib in rocm_lib_basenames:
+                            # If the dependency path is not within the ROCm library path, raise a warning
+                            # Check if dep_lib_path starts with rocm_lib_path(/opt/rocm/lib/) or real_rocm_lib_path(/opt/rocm-7.0.0/lib/) without symlink.
+                            if not (dep_lib_path.startswith(rocm_lib_path) or dep_lib_path.startswith(real_rocm_lib_path)):
+                                self.logger.warning(f"!!! Library {dep_lib} is linked to {dep_lib_path} which is outside of ROCm library path {rocm_lib_path}.")
+                                path_warnings.append(f"Library {dep_lib} is linked to {dep_lib_path} which is outside of ROCm library path {rocm_lib_path}.")
+
+            if missing:
+                missing_deps[lib] = missing
+            if path_warnings:
+                wrong_path_warnings[lib] = path_warnings
+
+        return missing_deps, wrong_path_warnings
+
+    def test_check_kernel_parameters(self):
```

Comment:
For this function, since we are executing same thing again and again hence, make a dictionary of lists , that list should contain the command line , error message, success message ,  exception message  , logging messages  and then loop on that dictionary. 

That would be more readable and quickly relatable